populus/project.py

```diff
@@ -159,46 +162,60 @@ def user_config(self):
 
     @property
     def project_config(self):
-
         if self._project_config_cache is None:
             project_config = Config(
                 config=self._project_config,
-                schema=self._project_config_schema,
             )
             project_config.unref()
             self._project_config_cache = project_config
 
+            # schema validation
+            # partial project config must be merged to get the entire schema
+            if self._user_config_cache is None:
+                self.user_config
+            self._merge_configs_and_validate_schema()
+
         return self._project_config_cache
 
+    def _merge_configs_and_validate_schema(self):
+        if self._merged_config_cache is None:
+            merged_config = copy.deepcopy(self._user_config_cache)
+            for key, value in self._project_config_cache.items(flatten=True):
+                if key != 'version':
+                    merged_config[key] = value
+            self._merged_config_cache = merged_config
+            # schema validation
+            Config(config=merged_config, schema=self._config_schema)
+
+        return self._merged_config_cache
+
     @property
     def config(self):
-
         warn_msg = '''Support for project.config will be dropped in the next populus release,
         use project_config or user_config or both'''
         warnings.warn(warn_msg, DeprecationWarning)
 
         if self._config_cache is None:
-
-            merged_config = copy.deepcopy(self.user_config)
-            for key, value in self.project_config.items(flatten=True):
-                if key != 'version':
-                    merged_config[key] = value
-            self._config_cache = merged_config
+            self.project_config
```

Comment:
I assume you are accessing these properties to populate their caches.  That stinks of *code-smell.  I would remove the `_merge_configs_and_validate_schema()` function and write a standalone function for merging configurations and pass it the `self.user_config` and `self.project_config`.

```python
@property
def config(self):
    if not self._merged_config_cache:
        self._merged_config_cache = merge_user_and_project_configs(self.user_config, self.project_config)
    return self._merged_config_cache
```

Reply:
absolutely